like many other "seq" analysis single cell rna-seq (ScRNA-seq) analysis also begins with a quality check on the initial data. Low quality libraries (cell) in a ScRNA-seq can srise from a veriety of sources.
Such as,
1. Cell damage during dissociation.
2. Failure in library preparation (inefficient reverse transcription or PCR amplification)
3. Inclusion of libraries with "low count" e.g., lot of cells with few expressed genes and high ```mitochondial``` (mt) and ```spike-in``` proportions.

We can apply few important quality metrices to indentify ```low quality cells```. Namely,
1. Library size (sum of all features [genes] in a cell). If it is low then it is considered a bad library.
2. Numbers of expressed features [# of genes with non-zero count]. If it is low then it is considered a bad library.
3. Proportion of reads mapped to ```spiked-in transcripts```. Since we start with equal numbers of transcripts added to each cell, any enrichment in ```spiked in transcripts``` shows that endogeneous genes has been lost and may results in poor library.
4. In the absense of ```spiked in transcripts``` proportion of reads mapped to ```mitochondrial``` genome can be an indication of the quality of a library. High proportion of reads mapped to ```mitochondrial``` genome indicates the low quality of library.

So above mentioned are the metrices we usually apply to assess the quality of a library and to calculate the QC metrices we use ```percellQCMetrics``` function from **scater** package.

But first we need to obtain some scRNA data to begin with. We will use data set from Lun et al. (2017) which is part of **scRNAseq** package

```r
library(scRNAseq)
sce.416b <- LunSpikeInData(which="416b") 
sce.416b$block <- factor(sce.416b$block)
```

We can look at the contents of the ```SingleCellExperiment``` object ```sce.416b```  

```r
> sce.416b
class: SingleCellExperiment 
dim: 46604 192 
metadata(0):
assays(1): counts
rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ...
  ENSMUSG00000095742 CBFB-MYH11-mcherry
rowData names(1): Length
colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1
  SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...
  SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1
  SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1
colData names(9): Source Name cell line ... spike-in addition block
reducedDimNames(0):
mainExpName: endogenous
altExpNames(2): ERCC SIRV
```

We can now use ```percellQCMetrics``` from **scater** package to calculate the quality metrices. This function gives out a DataFrame where rows represent cells (it can be noted here cells makes columns in SCE objects) and column represents few QC metrices.
These are,
**sum** column contains total sum of all genes.\
**detected** column contains number of expressed genes (i.e., all non-zero expressed genes).\
**subset_mito_percent** column contains percent of reads mapped to ```mt``` transcripts.\
**altexps_ERCC_percent** column contains percentage of reads mapped to ERCC transcripts.

