like many other "seq" analysis single cell rna-seq (ScRNA-seq) analysis also begins with a quality check on the initial data. Low quality libraries (cell) in a ScRNA-seq can srise from a veriety of sources.
Such as,
1. Cell damage during dissociation.
2. Failure in library preparation (inefficient reverse transcription or PCR amplification)
3. Inclusion of libraries with "low count" e.g., lot of cells with few expressed genes and high ```mitochondial``` and ```spike-in``` proportions.

We can apply few important quality metrices to indentify ```low quality cells```. Namely,
1. Library size (sum of all features [genes] in a cell). If it is low then it is considered a bad library.
2. Numbers of expressed features [# of genes with non-zero count]. If it is low then it is considered a bad library.
3. Proportion of reads mapped to ```spiked-in transcripts```. Since we start with equal numbers of transcripts added to each cell, any enrichment in ```spiked in transcripts``` shows that endogeneous genes has been lost and may results in poor library.
4. In the absense of ```spiked in transcripts``` proportion of reads mapped to ```mitochondrial``` genome can be an indication of the quality of a library. High proportion of reads mapped to ```mitochondrial``` genome indicates the low quality of library.

So above mentioned are the metrices we usually apply to assess the quality of a library and to calculate the QC metrices we use ```percellQCMetrics``` function from **scater** package.

But first we need to obtain some scRNA data to begin with. We will use data set from Lun et al. (2017) and is part of **scRNAseq** package
