More Often than not there is a systematic diffrence in sequencing coverage between libraries. They typically arise from technical differences in cDNA capture or PCR amplification efficiency across cells, and due to the difficulty of achieving consistent library preparation with minimal starting material. To remove these differences, Normalization is done. It aims to removes these differences such that they do not interfere with the comparison between expression profiles of cells. This ensures that the observed heterogeniety or diffrences in expression profiles is due to Biology and not because of the technical biases.

Most comman strategy is to use scaling Normalization. It is simplest and used commanly ormalization strategies. This involves dividing all counts for each cell by a cell-specific scaling factor, often called a “size factor”. The assumption here is that any cell-specific bias (e.g., in capture or amplification efficiency) affects all genes equally via scaling of the expected mean count for that cell. The size factor for each cell represents the estimate of the relative bias in that cell, so division of its counts by its size factor should remove that bias. The resulting Normalized expression count then can be used for further down stream analysis.

```r
> sce.zeisel <- ZeiselBrainData()
> sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
+                                       id=sub("_loc[0-9]+$", "", 
+                                              rownames(sce.zeisel)))

> rowData(sce.zeisel)$Ensembl <- mapIds(org.Mm.eg.db, 
+                                       keys=rownames(sce.zeisel), 
+                                       keytype="SYMBOL", column="ENSEMBL")

> stats <- perCellQCMetrics(sce.zeisel, subsets=list(
+   Mt=rowData(sce.zeisel)$featureType=="mito"))

> qc <- quickPerCellQC(stats, percent_subsets=c("altexps_ERCC_percent", 
+                                               "subsets_Mt_percent"))
> sce.zeisel <- sce.zeisel[,!qc$discard]

```
So now that we have obtained the data, aggregated it, mapped the transcripts (got the ensembl ID) with the appropriate database, calculated QC metrices for cells and finally obtained a `singleCellExperiment` object without the low quality cells (all the steps have been explained in previous quallity  control page), we are ready for the Normalization steps.


The “library size factor” for each cell is directly proportional to its library size where the proportionality constant is defined such that the mean size factor across all cells is equal to 1. let $sf_1$ be the size factor for $lib_1$ and $k_1$ be its proportionality constant. Let $N_{cell}$ be the total number of cells. Then according to the definition of the library size factor, $sf$, mean $sf$ should be equal to $1$ (i.e., $\bar {sf} = 1$). In other words,

```math
\bar{sf} = \frac{k_1(lib_1)+k_2(lib_2) + ....+ k_N(lib_N)}{N_{cell}} = 1
```

```r
> lib.sf.zeisel <- librarySizeFactors(sce.zeisel)

> sum(lib.sf.zeisel)
[1] 2816
```

`lib.size.factor` is a numeric vector such that the $\sum {sf} = N_{cell} $. In the Zeisel brain data, the library size factors differ by up to 10-fold across cells (see figure below). This is typical of the variability in coverage in scRNA-seq data.

```r
> hist(log10(lib.sf.zeisel), xlab="Log10[Size factor]", col='grey80')
```
![image3](https://user-images.githubusercontent.com/85447250/209002867-54df307f-6146-41d3-81c4-6bd1bc76b1e8.png)

It is important to note that computing `library size factor` depends upon following assumption. If a subset of gene is `DE` between cells `A` and `B`, it is assummed that there is no `imbalance` between `DE` genes between the said pairs. What we mean by no `imbalance` is , say in `A`,  the upregulation of a subset of genes is cancel out by the downregulation in  different subsets of genes. This ensures that the library size is an unbiased estimate of the relative cell-specific bias by avoiding composition effects. However, balanced DE is not generally present in scRNA-seq applications, which means that library size normalization may not yield accurate normalized expression values for downstream analyses.

In practice, normalization accuracy is not a major consideration for exploratory scRNA-seq data analyses. Composition biases do not usually affect the separation of clusters, only the magnitude - and to a lesser extent, direction - of the log-fold changes between clusters or cell types. As such, library size normalization is usually sufficient in many applications where the aim is to identify clusters and the top markers that define each cluster.







