More Often than not there is a systematic diffrence in sequencing coverage between libraries. They typically arise from technical differences in cDNA capture or PCR amplification efficiency across cells, and due to the difficulty of achieving consistent library preparation with minimal starting material. To remove these differences, Normalization is done. It aims to removes these differences such that they do not interfere with the comparison between expression profiles of cells. This ensures that the observed heterogeniety or diffrences in expression profiles is due to Biology and not because of the technical biases.

Most comman strategy is to use scaling Normalization. It is simplest and used commanly ormalization strategies. This involves dividing all counts for each cell by a cell-specific scaling factor, often called a “size factor”. The assumption here is that any cell-specific bias (e.g., in capture or amplification efficiency) affects all genes equally via scaling of the expected mean count for that cell. The size factor for each cell represents the estimate of the relative bias in that cell, so division of its counts by its size factor should remove that bias. The resulting Normalized expression count then can be used for further down stream analysis.

```r
> sce.zeisel <- ZeiselBrainData()
> sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
+                                       id=sub("_loc[0-9]+$", "", 
+                                              rownames(sce.zeisel)))

> rowData(sce.zeisel)$Ensembl <- mapIds(org.Mm.eg.db, 
+                                       keys=rownames(sce.zeisel), 
+                                       keytype="SYMBOL", column="ENSEMBL")

> stats <- perCellQCMetrics(sce.zeisel, subsets=list(
+   Mt=rowData(sce.zeisel)$featureType=="mito"))

> qc <- quickPerCellQC(stats, percent_subsets=c("altexps_ERCC_percent", 
+                                               "subsets_Mt_percent"))
> sce.zeisel <- sce.zeisel[,!qc$discard]

```
So now that we have obtained the data, aggregated it, mapped the transcripts (got the ensembl ID) with the appropriate database, calculated QC metrices for cells and finally obtained a `singleCellExperiment` object without the low quality cells (all the steps have been explained in previous quallity  control page), we are ready for the Normalization steps.



