More Often than not there is a systematic diffrence in sequencing coverage between libraries. They typically arise from technical differences in cDNA capture or PCR amplification efficiency across cells, and due to the difficulty of achieving consistent library preparation with minimal starting material. To remove these differences, Normalization is done. It aims to removes these differences such that they do not interfere with the comparison between expression profiles of cells. This ensures that the observed heterogeniety or diffrences in expression profiles is due to Biology and not because of the technical biases.

Most comman strategy is to use scaling Normalization. It is simplest and used commanly ormalization strategies. This involves dividing all counts for each cell by a cell-specific scaling factor, often called a “size factor”. The assumption here is that any cell-specific bias (e.g., in capture or amplification efficiency) affects all genes equally via scaling of the expected mean count for that cell. The size factor for each cell represents the estimate of the relative bias in that cell, so division of its counts by its size factor should remove that bias. The resulting Normalized expression count then can be used for further down stream analysis.

```r
> sce.zeisel <- ZeiselBrainData()
> sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
+                                       id=sub("_loc[0-9]+$", "", 
+                                              rownames(sce.zeisel)))

> rowData(sce.zeisel)$Ensembl <- mapIds(org.Mm.eg.db, 
+                                       keys=rownames(sce.zeisel), 
+                                       keytype="SYMBOL", column="ENSEMBL")

> stats <- perCellQCMetrics(sce.zeisel, subsets=list(
+   Mt=rowData(sce.zeisel)$featureType=="mito"))

> qc <- quickPerCellQC(stats, percent_subsets=c("altexps_ERCC_percent", 
+                                               "subsets_Mt_percent"))
> sce.zeisel <- sce.zeisel[,!qc$discard]

```
So now that we have obtained the data, aggregated it, mapped the transcripts (got the ensembl ID) with the appropriate database, calculated QC metrices for cells and finally obtained a `singleCellExperiment` object without the low quality cells (all the steps have been explained in previous quallity  control page), we are ready for the Normalization steps. 

### Library size Normalization

The “library size factor” for each cell is directly proportional to its library size where the proportionality constant is defined such that the mean size factor across all cells is equal to 1. let $sf_1$ be the size factor for $lib_1$ and $k_1$ be its proportionality constant. Let $N_{cell}$ be the total number of cells. Then according to the definition of the library size factor, $sf$, mean $sf$ should be equal to $1$ (i.e., $\bar {sf} = 1$). In other words,

```math
\bar{sf} = \frac{k_1(lib_1)+k_2(lib_2) + ....+ k_N(lib_N)}{N_{cell}} = 1
```

```r
> lib.sf.zeisel <- librarySizeFactors(sce.zeisel)

> sum(lib.sf.zeisel)
[1] 2816
```

`lib.size.factor` is a numeric vector such that the $\sum {sf} = N_{cell} $. In the Zeisel brain data, the library size factors differ by up to 10-fold across cells (see figure below). This is typical of the variability in coverage in scRNA-seq data.

```r
> hist(log10(lib.sf.zeisel), xlab="Log10[Size factor]", col='grey80')
```
![image3](https://user-images.githubusercontent.com/85447250/209002867-54df307f-6146-41d3-81c4-6bd1bc76b1e8.png)

It is important to note that computing `library size factor` depends upon following assumption. If a subset of gene is `DE` between cells A and B, it is assummed that there is no `imbalance` between `DE` genes between the said pairs. What we mean by no `imbalance` is , say in A,  the upregulation of a subset of genes is cancel out by the downregulation in  different subsets of genes. This ensures that the library size is an unbiased estimate of the relative cell-specific bias by avoiding composition effects. However, balanced DE is not generally present in scRNA-seq applications, which means that library size normalization may not yield accurate normalized expression values for downstream analyses.

In practice, normalization accuracy is not a major consideration for exploratory scRNA-seq data analyses. Composition biases do not usually affect the separation of clusters, only the magnitude - and to a lesser extent, direction - of the log-fold changes between clusters or cell types. As such, library size normalization is usually sufficient in many applications where the aim is to identify clusters and the top markers that define each cluster.

### Normalization by deconvolution

In the preceding section we have said that `imbalances` in `DE` of genes between cells indicates composition biases. Consider the simple example of two cells where a single gene $X$ is upregulated in one cell $A$ compared to the other cell $B$. This upregulation means that either (1) more sequencing resources are devoted to $X$ in $A$, thus decreasing coverage of all other non-DE genes when the total library size of each cell is experimentally fixed (e.g., due to library quantification); or (2) the library size of $A$ increases when $X$ is assigned more reads or UMIs, increasing the library size factor and yielding smaller normalized expression values for all non-DE genes. In both cases, the net effect is that non-DE genes in $A$ will incorrectly appear to be downregulated compared to $B$.

Since we have a lot zeros and low counts on scRNA data, we pool counts from many cells to increase the size of the counts for accurate size factor estimation. Pool-based size factors are then "deconvolved" into cell-based factors for normalization of each cell’s expression profile. This is performed using the `calculateSumFactors()` function from **scran**.

```r
set.seed(100)
clust.zeisel <- quickCluster(sce.zeisel)

> class(clust.zeisel)
[1] "factor"

> head(clust.zeisel, 27)
 [1] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 4 2 2 2 2 4 2 2
Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14

> table(clust.zeisel)
clust.zeisel
  1   2   3   4   5   6   7   8   9  10  11  12  13  14 
170 254 441 178 393 148 219 240 189 123 112 103 135 111
```
**explanation**

We use `quickCluster()` to aggregate/condense the count matrix, because most of the genes in most of the cells are either not expressed at all ( 0 in count matrix) or are having an extremely low counts. So this function "aggregate" all the counts to a "pooled count matrix" which is further used for "library size factor" calculation. `quickCluster()` results in clustering cells based upon some kind of "dissimilarity" between any two cells (in easy words it just separates cells and put it in different bins). `quickCluster()` produces a `factor` with diffrent `levels` where each cells belong to a different level

Next, to calculte the `library size factors` for each cells one can use `calculateSumFactors()`. what it does is that it calculates Pool-based size factors from "pooled count matrix" and then “deconvolved” into cell-based factors for normalization of each cell’s expression profile, as shown below.

```r
>deconv.sf.zeisel <- calculateSumFactors(sce.zeisel, cluster=clust.zeisel)
```

So, the logical flow of commands will be `quickCluster(original SCE)` $>$ `calculateSumfactors(original SCE, cluster=pooled count matrix obtained from quickCluster())` $>$ gives `deconvolve` size factors.

Now we can see the compatibility between "lib.sf.zeisel" and "deconv.sf.zeisel" by plotting one aginst other.

```r
plot(lib.sf.zeisel, deconv.sf.zeisel, xlab="Library size factor",
    ylab="Deconvolution size factor", log='xy', pch=16,
    col=as.integer(factor(sce.zeisel$level1class))))
 ```
 
 ![image4](https://user-images.githubusercontent.com/85447250/209418955-f5fbf3cd-7332-4206-9353-ca860e8f8bd9.png)
 
 
 ### Scaling and log-transforming
 Once we have computed the size factors, we use the `logNormCounts()` function from **scater** to compute normalized expression values for each cell. This is done by dividing the count for each gene/spike-in transcript with the appropriate size factor for that cell. The function also log-transforms the normalized values, creating a new assay called "logcounts". (Technically, these are “log-transformed normalized expression values”). These log-values will be the basis of our downstream analyses. 
 
 ```r
 > set.seed(100)
> clust.zeisel <- quickCluster(sce.zeisel) 
> sce.zeisel <- computeSumFactors(sce.zeisel, cluster=clust.zeisel, min.mean=0.1)
> sce.zeisel <- logNormCounts(sce.zeisel)
```

**explanation**

Difference between `calculateSumfactors()` and `computeSumfactors()` is that the former returns a "numeric vector" containing sizeFactors of `SingleCellExperiment` object, while the later one would return an `SingleCellExperiment` object with size factors contained in `colData` "sizeFactor". 

```r
> head(colData(sce.zeisel))
DataFrame with 6 rows and 11 columns
                    tissue   group # total mRNA mol      well       sex       age
               <character> <numeric>      <numeric> <numeric> <numeric> <numeric>
1772071015_C02    sscortex         1           1221         3         3         2
1772071017_G12    sscortex         1           1231        95         1         1
1772071017_A05    sscortex         1           1652        27         1         1
1772071014_B06    sscortex         1           1696        37         3         2
1772067065_H06    sscortex         1           1219        43         3         6
1772071017_E02    sscortex         1           1378         5         1         1
                diameter        cell_id  level1class level2class sizeFactor
               <numeric>    <character>  <character> <character>  <numeric>
1772071015_C02         1 1772071015_C02 interneurons       Int10    1.53302
1772071017_G12       353 1772071017_G12 interneurons       Int10    1.51978
1772071017_A05        13 1772071017_A05 interneurons        Int6    2.31771
1772071014_B06        19 1772071014_B06 interneurons       Int10    2.30982
1772067065_H06        12 1772067065_H06 interneurons        Int9    1.63565
1772071017_E02        21 1772071017_E02 interneurons        Int9    1.91402

```



 
 
 

 













