### Assigning cell label from reference data

+ Using existing reference

We will use method from `SingleR`. This method assign label to an unknown cell (our test data) from a reference samples based on the spearman rank correlation between the marker genes of our unknown cell and the reference sample. But first we need to get out test data and a reference sample. We will agin use 10x PBMC as test data and use `celldex` (a library pacage which contains curated data sets) as a source of reference sample.

```r

# Loading the data

raw.path <- getTestFile("tenx-2.1.0-pbmc4k/1.0.0/raw.tar.gz")
out.path <- file.path(tempdir(), "pbmc4k")
untar(raw.path, exdir=out.path)

fname <- file.path(out.path, "raw_gene_bc_matrices/GRCh38")
sce.pbmc <- read10xCounts(fname, col.names=TRUE)
colnames(rowData(sce.pbmc))

# Gene annotation

rownames(sce.pbmc) <- uniquifyFeatureNames(rowData(sce.pbmc)$ID, 
                                           rowData(sce.pbmc)$Symbol)
location <- mapIds(EnsDb.Hsapiens.v86, keys = rowData(sce.pbmc)$ID,
                   keytype = "GENEID", column = "SEQNAME")

# Cell detection
set.seed(1000)
e.out <- emptyDrops(counts(sce.pbmc))
sce.pbmc <- sce.pbmc[,which(e.out$FDR <= 0.001)]

# Quality control
stats <- perCellQCMetrics(sce.pbmc, subsets=list(Mito=which(location=="MT")))
high.mito <- isOutlier(stats$subsets_Mito_percent, type="higher")
sce.pbmc <- sce.pbmc[,!high.mito]

# Normalization 

set.seed(100)
clusters <- quickCluster(sce.pbmc)
sce.pbmc <- computeSumFactors(sce.pbmc, cluster=clusters)
sce.pbmc <- logNormCounts(sce.pbmc)

# Variance modeling

set.seed(1001)
dec.pbmc <- modelGeneVarByPoisson(sce.pbmc)
top.pbmc <- getTopHVGs(dec.pbmc, prop=0.1)

# Dimensinality reduction 
g <- buildSNNGraph(sce.pbmc, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
colLabels(sce.pbmc) <- factor(clust)

> sce.pbmc
class: SingleCellExperiment 
dim: 33694 3994 
metadata(1): Samples
assays(2): counts logcounts
rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
rowData names(2): ID Symbol
colnames(3994): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
  TTTGTCAGTTAAGACA-1 TTTGTCATCCCAAGAT-1
colData names(4): Sample Barcode sizeFactor label
reducedDimNames(3): PCA TSNE UMAP
mainExpName: NULL
altExpNames(0):
```

Now we will create a reference with BlueprintEncodeData() function from `celldex` package.
This function provides normalized expression values for 259 bulk RNA-seq samples generated by Blueprint and ENCODE projects (Martens and Stunnenberg, 2013; The ENCODE Consortium, 2012). BlueprintEncodeData() function will return a 'SummarizedExperiment' object with a "logcounts" assay containing the log-normalized expression values, along with cell type labels in the colData. Blueprint Epigenomics contains 144 RNA-seq pure immune samples annotated to 28 cell types. ENCODE contains 115 RNA-seq pure stroma and immune samples annotated to 17 cell types. All together, this reference contains 259 samples with 43 cell types ("label.fine"), manually aggregated into 25 broad classes (see below tabulated summary for "label.main"). 

```r
ref <- BlueprintEncodeData()

> ref
class: SummarizedExperiment 
dim: 19859 259 
metadata(0):
assays(1): logcounts
rownames(19859): TSPAN6 TNMD ... LINC00550 GIMAP1-GIMAP5
rowData names(0):
colnames(259): mature.neutrophil
  CD14.positive..CD16.negative.classical.monocyte ...
  epithelial.cell.of.umbilical.artery.1
  dermis.lymphatic.vessel.endothelial.cell.1
colData names(3): label.main label.fine label.ont

> table(colData(ref)$label.main)

       Adipocytes        Astrocytes           B-cells      CD4+ T-cells 
                9                 2                 8                14 
     CD8+ T-cells      Chondrocytes                DC Endothelial cells 
                5                 2                 1                26 
      Eosinophils  Epithelial cells      Erythrocytes       Fibroblasts 
                1                18                 7                20 
              HSC     Keratinocytes       Macrophages       Melanocytes 
               38                 2                25                 4 
  Mesangial cells         Monocytes          Myocytes           Neurons 
                2                16                 4                 4 
      Neutrophils          NK cells         Pericytes   Skeletal muscle 
               23                 3                 2                 7 
    Smooth muscle 
               16 
  
  > head(factor(colData(ref)$label.main))
[1] Neutrophils Monocytes   Neutrophils HSC         Neutrophils Monocytes  
25 Levels: Adipocytes Astrocytes B-cells CD4+ T-cells CD8+ T-cells ... Smooth muscle
```



   







