### Assigning cell label from reference data

+ Using existing reference

We will use method from `SingleR`. This method assign label to an unknown cell (our test data) from a reference samples based on the spearman rank correlation between the marker genes of our unknown cell and the reference sample. But first we need to get out test data and a reference sample. We will agin use 10x PBMC as test data and use `celldex` (a library pacage which contains curated data sets) as a source of reference sample.

```r

# Loading the data

raw.path <- getTestFile("tenx-2.1.0-pbmc4k/1.0.0/raw.tar.gz")
out.path <- file.path(tempdir(), "pbmc4k")
untar(raw.path, exdir=out.path)

fname <- file.path(out.path, "raw_gene_bc_matrices/GRCh38")
sce.pbmc <- read10xCounts(fname, col.names=TRUE)
colnames(rowData(sce.pbmc))

# Gene annotation

rownames(sce.pbmc) <- uniquifyFeatureNames(rowData(sce.pbmc)$ID, 
                                           rowData(sce.pbmc)$Symbol)
location <- mapIds(EnsDb.Hsapiens.v86, keys = rowData(sce.pbmc)$ID,
                   keytype = "GENEID", column = "SEQNAME")

# Cell detection
set.seed(1000)
e.out <- emptyDrops(counts(sce.pbmc))
sce.pbmc <- sce.pbmc[,which(e.out$FDR <= 0.001)]

# Quality control
stats <- perCellQCMetrics(sce.pbmc, subsets=list(Mito=which(location=="MT")))
high.mito <- isOutlier(stats$subsets_Mito_percent, type="higher")
sce.pbmc <- sce.pbmc[,!high.mito]

# Normalization 

set.seed(100)
clusters <- quickCluster(sce.pbmc)
sce.pbmc <- computeSumFactors(sce.pbmc, cluster=clusters)
sce.pbmc <- logNormCounts(sce.pbmc)

# Variance modeling

set.seed(1001)
dec.pbmc <- modelGeneVarByPoisson(sce.pbmc)
top.pbmc <- getTopHVGs(dec.pbmc, prop=0.1)

# Dimensinality reduction 
g <- buildSNNGraph(sce.pbmc, k=10, use.dimred = 'PCA')
clust <- igraph::cluster_walktrap(g)$membership
colLabels(sce.pbmc) <- factor(clust)

> sce.pbmc
class: SingleCellExperiment 
dim: 33694 3994 
metadata(1): Samples
assays(2): counts logcounts
rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
rowData names(2): ID Symbol
colnames(3994): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
  TTTGTCAGTTAAGACA-1 TTTGTCATCCCAAGAT-1
colData names(4): Sample Barcode sizeFactor label
reducedDimNames(3): PCA TSNE UMAP
mainExpName: NULL
altExpNames(0):
```

Now we will create a reference with BlueprintEncodeData() function from `celldex` package.
This function provides normalized expression values for 259 bulk RNA-seq samples generated by Blueprint and ENCODE projects (Martens and Stunnenberg, 2013; The ENCODE Consortium, 2012). BlueprintEncodeData() function will return a 'SummarizedExperiment' object with a "logcounts" assay containing the log-normalized expression values, along with cell type labels in the colData. Blueprint Epigenomics contains 144 RNA-seq pure immune samples annotated to 28 cell types. ENCODE contains 115 RNA-seq pure stroma and immune samples annotated to 17 cell types. All together, this reference contains 259 samples with 43 cell types ("label.fine"), manually aggregated into 25 broad classes (see below tabulated summary for "label.main"). 

```r
ref <- BlueprintEncodeData()

> ref
class: SummarizedExperiment 
dim: 19859 259 
metadata(0):
assays(1): logcounts
rownames(19859): TSPAN6 TNMD ... LINC00550 GIMAP1-GIMAP5
rowData names(0):
colnames(259): mature.neutrophil
  CD14.positive..CD16.negative.classical.monocyte ...
  epithelial.cell.of.umbilical.artery.1
  dermis.lymphatic.vessel.endothelial.cell.1
colData names(3): label.main label.fine label.ont

> table(colData(ref)$label.main)

       Adipocytes        Astrocytes           B-cells      CD4+ T-cells 
                9                 2                 8                14 
     CD8+ T-cells      Chondrocytes                DC Endothelial cells 
                5                 2                 1                26 
      Eosinophils  Epithelial cells      Erythrocytes       Fibroblasts 
                1                18                 7                20 
              HSC     Keratinocytes       Macrophages       Melanocytes 
               38                 2                25                 4 
  Mesangial cells         Monocytes          Myocytes           Neurons 
                2                16                 4                 4 
      Neutrophils          NK cells         Pericytes   Skeletal muscle 
               23                 3                 2                 7 
    Smooth muscle 
               16 
  
  > head(factor(colData(ref)$label.main))
[1] Neutrophils Monocytes   Neutrophils HSC         Neutrophils Monocytes  
25 Levels: Adipocytes Astrocytes B-cells CD4+ T-cells CD8+ T-cells ... Smooth muscle
```

Now we are ready to use `SingleR()` function to annotate each of our PBMCs with the main cell type labels from the Blueprint/ENCODE reference. This returns a `DataFrame` where each row corresponds to a cell in the test dataset and contains its label assignments (see "label" column in the pred DataFrame). 

```r
pred <- SingleR(test=sce.pbmc, ref=ref, labels=ref$label.main)

head(pred)
DataFrame with 6 rows and 5 columns
                                           scores first.labels       tuning.scores
                                         <matrix>  <character>         <DataFrame>
AAACCTGAGAAGGCCT-1 0.251873:0.118890:0.287805:...    Monocytes 0.413717:0.00491616
AAACCTGAGACAGACC-1 0.280080:0.133100:0.334590:...    Monocytes 0.469530:0.40291032
AAACCTGAGGCATGGT-1 0.211503:0.153752:0.345878:... CD4+ T-cells 0.180486:0.06419868
AAACCTGCAAGGTTCT-1 0.218346:0.155343:0.367597:... CD8+ T-cells 0.307063:0.19412309
AAACCTGCAGGCGATA-1 0.323697:0.182205:0.483679:...    Monocytes 0.558153:0.50413520
AAACCTGCATGAAGTA-1 0.312739:0.166972:0.461094:...    Monocytes 0.531659:0.46923231
                         labels pruned.labels
                    <character>   <character>
AAACCTGAGAAGGCCT-1    Monocytes     Monocytes
AAACCTGAGACAGACC-1    Monocytes     Monocytes
AAACCTGAGGCATGGT-1 CD4+ T-cells  CD4+ T-cells
AAACCTGCAAGGTTCT-1 CD8+ T-cells  CD8+ T-cells
AAACCTGCAGGCGATA-1    Monocytes     Monocytes
AAACCTGCATGAAGTA-1    Monocytes     Monocytes

> table(pred$labels)

     B-cells CD4+ T-cells CD8+ T-cells           DC  Eosinophils Erythrocytes 
         548          773         1275            1            1            5 
         HSC    Monocytes     NK cells 
          14         1126          251 
> 
```

Once we are done with creating an annotation DataFrame (pred), we are ready to plot it with `plotScoreHeatmap()`

```r
plotScoreHeatmap(pred)
```

![image21](https://user-images.githubusercontent.com/85447250/213538493-b8698ed1-b1bd-4ca5-bcd0-e1a2c8ad1da7.png)

Fig. Heatmap of the assignment score for each cell (column) and label (row).

**explanation for the plot above**

columns represents the cells from our test `SingleCellExperiment` object (sce.pbmc). Rows represents the labels (types of cells ) from the main annotation `SingleCellExperiment` object (ref here; if we remember it is the reference sample conataining all the annotations from Blueprint and ENCODE projects). Here we have 25 rows (cell types from ref) and we can tally this row with `factor(ref$label.main)`. So the reading for the heatmap goes like this; for every test cell (column) what are the scores for different labels (rows)?. Ideally, each cell in our test data should be annotated/assigned/recognized with just one of the labels (rows). For example, for first few columns if we look at the bottom left corner of the plot we do see a high score (bright yellow color) for B-cells (see rows). So, this bunch of cells has been annotated with B-cells (and luckey for us that we don't see yellow for any other rows. If that was the case we would be in trouble annotating these "first few columns" because in that case we would have an additional row/label to chose from and that is not always an easy task). We also see "Labels" to the right of scoring bar (a vertical bar with color gradient spanning higher to lower). This "Labels" indicates the recognized/annotated cells found in our test data set.

**end of explanation**

Another excercise that we can do is to compare assignments of cells with the clustering results to find out the identity of clusters.

```r
tab <- table(Assigned=pred$pruned.labels, Cluster=colLabels(sce.pbmc))

```

**explanation of "tab"**
 
 ```r
 > head(pred$pruned.labels, 25)
 [1] "Monocytes"    "Monocytes"    "CD4+ T-cells" "CD8+ T-cells" "Monocytes"   
 [6] "Monocytes"    "B-cells"      "CD8+ T-cells" "CD4+ T-cells" "CD4+ T-cells"
[11] "Monocytes"    "CD8+ T-cells" "CD8+ T-cells" "CD8+ T-cells" "Monocytes"   
[16] "CD8+ T-cells" "CD8+ T-cells" "CD4+ T-cells" "B-cells"      "CD8+ T-cells"
[21] "Monocytes"    "Monocytes"    "CD8+ T-cells" "NK cells"     "Monocytes" 

> head(colLabels(sce.pbmc), 25)
 [1] 2  2  9  3  6  1  4  9  9  3  6  3  9  11 6  9  3  3  4  9  2  10 9  8  2 
Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16
```
So, the formation of the table "tab" goes like this. We do have two sets of data (pred$pruned.labels and colLabels(sce.pbmc)) of equal length (=length(colnames(sce.pbmc))). We look for "Monocytes" in first row of head(pred$pruned.labels, 25). We find first and second indexes have "Monocytes" on them. Now we look at the first five indexes in head(colLabels(sce.pbmc), 25) we see first two indexes are labeled "2" (these two cells belong to cluster "2"). So what information do we get from here? it tells us that among the first five cells two are "Monocytes" and belong to cluster "2". We can extend this definition to find out what cluster contains which cell type in a tabular format (as seen below). 

```r
> tab
              Cluster
Assigned         1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16
  B-cells        0   0   0 528   0   0   0   0   0  15   0   0   0   0   0   0
  CD4+ T-cells   1   0 262   0   1   0   0   0 442   0   0   0   0  58   0   2
  CD8+ T-cells   0   1 327   8 318   0   0   0 412   0 133   0   4  63   0   2
  DC             0   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0
  Eosinophils    0   0   0   0   0   0   1   0   0   0   0   0   0   0   0   0
  Erythrocytes   0   0   1   0   0   0   4   0   0   0   0   0   0   0   0   0
  HSC            0   0   0   0   0   0   0   0   0  14   0   0   0   0   0   0
  Monocytes     55 729   0   0   0 125   0   0   0  16   0  69   0   0  84   0
  NK cells       0   0   0   2  32   0   0 166   0   0  16   0  26   0   0   1
  ```

 
 
 
 




   







